File: leetcode/minimum-deletions-to-make-string-balanced.py
class Solution(object):
    def minimumDeletions(self, s):
        b_count = 0
        min_deletions = 0

        for char in s:
            if char == 'b':
                b_count += 1

            else:
                min_deletions = min(min_deletions + 1, b_count)
        
        return min_deletions

if __name__ == "__main__":
    solver = Solution()
    
    def run_test(case_num, s, expected):
        result = solver.minimumDeletions(s)
        status = "PASS" if result == expected else "FAIL"
        print(f"Test Case {case_num}: '{s}' -> {result} (Expected: {expected}) -> {status}")

    # Example 1
    # We can delete 'b' at index 2 and 'b' at index 6 -> "aaabbb" (2 deletions)
    # OR delete 'a' at index 3 and 'a' at index 6 -> "aabbbb" (2 deletions)
    run_test(1, "aababbab", 2)

    # Example 2
    # Delete the two leading 'b's -> "aaaaabb"
    run_test(2, "bbaaaaabb", 2)
    
    # Custom Case: Already balanced
    run_test(3, "aaabbb", 0)

    # Custom Case: All 'b's then 'a's (Worst case)
    # "bbbaaa" -> delete 3 'b's OR 3 'a's -> cost 3
    run_test(4, "bbbaaa", 3)

File: golang/minimum-deletions-to-make-string-balanced/minimum-deletions-to-make-string-balanced.go
package main

import (
	"fmt"
	"reflect"
)

func minimumDeletions(s string) int {
	bCount := 0
	minDeletions := 0

	for _, char := range s {
		if char == 'b' {
			bCount++
		} else {
			if minDeletions+1 < bCount {
				minDeletions = minDeletions + 1
			} else {
				minDeletions = bCount
			}
		}
	}
	return minDeletions
}

func runTest(testNum int, s string, expected int) {
	result := minimumDeletions(s)

	status := "FAIL"
	if reflect.DeepEqual(result, expected) {
		status = "PASS"
	}

	fmt.Printf(
		"Test Case %d: '%s' -> %d (Expected: %d) -> %s\n",
		testNum, s, result, expected, status,
	)
}

func main() {
	runTest(1, "aababbab", 2)
	runTest(2, "bbaaaaabb", 2)
	runTest(3, "aaabbb", 0)
	runTest(4, "bbbaaa", 3)
}
